clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerLoginModule.java

```diff
@@ -322,60 +343,80 @@ private void identifyExtensions() throws LoginException {
+            for (Iterator<Object> iterator = subject.getPrivateCredentials().iterator(); iterator.hasNext(); ) {
+                Object privateCredential = iterator.next();
+                if (privateCredential == myCommittedToken) {
+                    iterator.remove();
+                    myCommittedToken = null;
+                    break;
+                }
             }
+            log.info("Done logging out my token; committed token count is now {}", committedTokenCount());
+        } else {
+            log.info("No tokens to logout for this login");
         }
-        log.info("Done logging out my token; committed token count is now {}", committedTokenCount());
 
-        log.info("Logging out my extensions");
-        if (subject.getPublicCredentials().removeIf(e -> myCommittedExtensions == e))
-            myCommittedExtensions = null;
-        log.info("Done logging out my extensions");
+        if (myCommittedExtensions != null) {
+            log.info("Logging out my extensions");
+            if (subject.getPublicCredentials().removeIf(e -> myCommittedExtensions == e))
+                myCommittedExtensions = null;
+            log.info("Done logging out my extensions");
+        } else {
+            log.info("No extensions to logout for this login");
+        }
 
+        loginState = LoginState.NOT_LOGGED_IN;
         return true;
     }
 
     @Override
     public boolean commit() {
-        if (tokenRequiringCommit == null) {
+        if (loginState != LoginState.LOGGED_IN_NOT_COMMITTED) {
             if (log.isDebugEnabled())
                 log.debug("Nothing here to commit");
             return false;
         }
 
-        log.info("Committing my token; current committed token count = {}", committedTokenCount());
-        subject.getPrivateCredentials().add(tokenRequiringCommit);
-        myCommittedToken = tokenRequiringCommit;
-        tokenRequiringCommit = null;
-        log.info("Done committing my token; committed token count is now {}", committedTokenCount());
+        if (tokenRequiringCommit != null) {
+            log.info("Committing my token; current committed token count = {}", committedTokenCount());
+            subject.getPrivateCredentials().add(tokenRequiringCommit);
+            myCommittedToken = tokenRequiringCommit;
+            tokenRequiringCommit = null;
+            log.info("Done committing my token; committed token count is now {}", committedTokenCount());
+        } else {
+            log.info("No tokens to commit, this login cannot be used to establish client connections");
+        }
 
-        subject.getPublicCredentials().add(extensionsRequiringCommit);
-        myCommittedExtensions = extensionsRequiringCommit;
-        extensionsRequiringCommit = null;
+        if (extensionsRequiringCommit != null) {
+            if (myCommittedToken == null)
+                throw new IllegalStateException("Extensions being committed without a token");
```

Comment:
Is the `myCommittedToken == null` check unnecessary here since it can never be the case when there are extensions?  I think we make sure of this since we only call `identifyExtensions()` when there is a token.